README.md
# vue-web-extension

This template should help get you started developing with Vue 3 in Vite.

## Recommended IDE Setup

[VSCode](https://code.visualstudio.com/) + [Volar](https://marketplace.visualstudio.com/items?itemName=Vue.volar) (and disable Vetur) + [TypeScript Vue Plugin (Volar)](https://marketplace.visualstudio.com/items?itemName=Vue.vscode-typescript-vue-plugin).

## Customize configuration

See [Vite Configuration Reference](https://vitejs.dev/config/).

## Project Setup

```sh
npm install
```

### Compile and Hot-Reload for Development

```sh
npm run dev
```

### Compile and Minify for Production

```sh
npm run build
```

### Run Unit Tests with [Vitest](https://vitest.dev/)

```sh
npm run test:unit
```

### Run End-to-End Tests with [Cypress](https://www.cypress.io/)

```sh
npm run build
npm run test:e2e # or `npm run test:e2e:ci` for headless testing
```

### Lint with [ESLint](https://eslint.org/)

```sh
npm run lint
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 2.0.1 (2022-08-09)



### [2.0.0](https://github.com/Kocal/vue-web-extension/compare/v1.12.0...v2.0.0) (2020-08-31)

Vue-web-extension now becomes a [Vue CLI 3+ preset](https://cli.vuejs.org/guide/plugins-and-presets.html#presets)!

Support for Vue CLI <2 has been removed.

See [#622](https://github.com/Kocal/vue-web-extension/pull/622) for more information!

### Bug Fixes

* images & fonts not being properly bundled ([#571](https://github.com/Kocal/vue-web-extension/issues/571)) ([91de1d5](https://github.com/Kocal/vue-web-extension/commit/91de1d5f031ce17d9b3c1ffeaf38e855d4251d23))

## [1.12.0](https://github.com/Kocal/vue-web-extension/compare/v1.11.0...v1.12.0) (2019-08-15)


### Bug Fixes

* change fonts directory to `/fonts/` ([#453](https://github.com/Kocal/vue-web-extension/issues/453)) ([8700217](https://github.com/Kocal/vue-web-extension/commit/8700217))
* fix images loading when using a relative path ([#474](https://github.com/Kocal/vue-web-extension/issues/474)) ([795d91a](https://github.com/Kocal/vue-web-extension/commit/795d91a))


### Features

* **webpack:** handle fonts ([#445](https://github.com/Kocal/vue-web-extension/issues/445)) ([9ee09ae](https://github.com/Kocal/vue-web-extension/commit/9ee09ae))



## [1.11.0](https://github.com/Kocal/vue-web-extension/compare/v1.10.0...v1.11.0) (2019-06-08)


### Build System

* **deps:** update ESLint ([688a2d5](https://github.com/Kocal/vue-web-extension/commit/688a2d5))
* **deps:** update Vue to ^2.6.10 ([df9ae97](https://github.com/Kocal/vue-web-extension/commit/df9ae97))


### Features

* install babel runtime ([#411](https://github.com/Kocal/vue-web-extension/issues/411)) ([4319a38](https://github.com/Kocal/vue-web-extension/commit/4319a38))
* use core-js 3 ([#410](https://github.com/Kocal/vue-web-extension/issues/410)) ([e27d910](https://github.com/Kocal/vue-web-extension/commit/e27d910))
* **hot-reload:** migrate to webpack-extension-reloader ([#431](https://github.com/Kocal/vue-web-extension/issues/431)) ([cd68341](https://github.com/Kocal/vue-web-extension/commit/cd68341))



<a name="1.10.0"></a>

# [1.10.0](https://github.com/Kocal/vue-web-extension/compare/v1.9.0...v1.10.0) (2019-04-05)

### Bug Fixes

- Handle `eval` removal properly, remove "remove-evals" script and WebpackShellPlugin ([#398](https://github.com/Kocal/vue-web-extension/issues/398)) ([ab1ce52](https://github.com/Kocal/vue-web-extension/commit/ab1ce52))

### Features

- add core-js dependency ([#381](https://github.com/Kocal/vue-web-extension/issues/381)) ([01066ba](https://github.com/Kocal/vue-web-extension/commit/01066ba))



<a name="1.9.0"></a>
# [1.9.0](https://github.com/Kocal/vue-web-extension/compare/v1.8.1...v1.9.0) (2018-11-08)


### Features

* `browser` object is now accessible in Vue templates by using `$browser` ([#357](https://github.com/Kocal/vue-web-extension/issues/357)) ([83b76fe](https://github.com/Kocal/vue-web-extension/commit/83b76fe))



<a name="1.8.1"></a>
## [1.8.1](https://github.com/Kocal/vue-web-extension/compare/v1.8.0...v1.8.1) (2018-10-26)


### Bug Fixes

* remove vulnerable "zip-folder" package for "archiver" ([#347](https://github.com/Kocal/vue-web-extension/issues/347)) ([a42d20f](https://github.com/Kocal/vue-web-extension/commit/a42d20f)), closes [#338](https://github.com/Kocal/vue-web-extension/issues/338)



<a name="1.8.0"></a>
# [1.8.0](https://github.com/Kocal/vue-web-extension/compare/v1.7.0...v1.8.0) (2018-10-23)


### Features

* use EJS on html files, inject process.env.* vars ([#336](https://github.com/Kocal/vue-web-extension/issues/336)) ([8171f3e](https://github.com/Kocal/vue-web-extension/commit/8171f3e)), closes [/github.com/Kocal/vue-web-extension/issues/334#issuecomment-430491330](https://github.com//github.com/Kocal/vue-web-extension/issues/334/issues/issuecomment-430491330)



<a name="1.7.0"></a>
# [1.7.0](https://github.com/Kocal/vue-web-extension/compare/v1.6.0...v1.7.0) (2018-10-14)


### Bug Fixes

* **ci:** use TRAVIS_PULL_REQUEST_BRANCH env var ([870bc9b](https://github.com/Kocal/vue-web-extension/commit/870bc9b))


### Features

* **prompt:** add prompt for Mozilla Polyfill ([#335](https://github.com/Kocal/vue-web-extension/issues/335)) ([ec0277b](https://github.com/Kocal/vue-web-extension/commit/ec0277b))



<a name="1.6.0"></a>
# [1.6.0](https://github.com/Kocal/vue-web-extension/compare/v1.5.0...v1.6.0) (2018-07-08)


### Features

* add prompt for options page ([#263](https://github.com/Kocal/vue-web-extension/issues/263)) ([0bdb5e6](https://github.com/Kocal/vue-web-extension/commit/0bdb5e6))



<a name="1.5.0"></a>
# [1.5.0](https://github.com/Kocal/vue-web-extension/compare/v1.4.0...v1.5.0) (2018-07-01)


### Bug Fixes

* prettier git hook is not asked during init ([#255](https://github.com/Kocal/vue-web-extension/issues/255)) ([391a17b](https://github.com/Kocal/vue-web-extension/commit/391a17b))

public/data.js
/**data.js
 * 데이터스크래핑 내역을 확인, 수정, 다음 액션 지정하는 부분
 * data.html에 <button>, <table> 형태를 만들어준 뒤
 * 들어오는 데이터에 따라 tr과 td를 append해 테이블을 그려준다.
 */

let confirmBtn = document.getElementsByTagName('BUTTON');
let tbody = document.getElementsByTagName('TBODY');

let xpathArr = []; //데이터추가 된 후 xpath 스크래핑 내역
let textArr = []; //데이터추가 된 후 innerText 스크래핑 내역
let typeArr = []; // 스크래핑된 엘리먼트 최상위 태그 이름
let originArr = []; // 데이터추가 되기 전 xpath 스크래핑 내역
let originTextArr = []; // 데이터추가 되기 전 innerText 스크래핑 내역
let secondElementArr = [];
let image;
let scrapeData =[];

//버튼에 리스터 추가 storage의 데이터를 가져옴
function getData(){
  console.log(confirmBtn);
  confirmBtn[0].addEventListener('click', add);
  confirmBtn[1].addEventListener('click', resetTable);
  confirmBtn[2].addEventListener('click', save);
  confirmBtn[3].addEventListener('click', nextPage);
  confirmBtn[4].addEventListener('click', addSecondEle);

  chrome_storage_lock = false;
      chrome.storage.local.get({
        scrapeData:[]}, function(result){
          scrapeData = result.scrapeData;
          textArr = scrapeData[0].textData;
          typeArr = scrapeData[0].typeArr;
          originArr = scrapeData[0].originData;
          originTextArr = scrapeData[0].originTextData;
          image = scrapeData[0].image;
          xpathArr = scrapeData[0].xpathData;
          makeTable(textArr);
      });
}
 
function makeTable(textArr){
  console.log(secondElementArr);
  console.log(typeArr);
  for(var i = 0 ; i< textArr.length; i++){
    let tr_01 = document.createElement('tr');
    if(typeof(textArr[i]) != 'string'){
      for(var j = 0 ; j< textArr[i].length; j++){
        let td_01 = document.createElement('td');
            td_01.textContent = textArr[i][j];
            // 툴팁 추가
            td_01.setAttribute('title', textArr[i][j]);
            tr_01.appendChild(td_01);
          }
        }else{
          let td_01 = document.createElement('td');
          td_01.textContent = textArr[i];
          // 툴팁 추가
          td_01.setAttribute('title', textArr[i]);
          tr_01.appendChild(td_01);
    }
    tbody[0].appendChild(tr_01);
  }
  // 테이블 생성 후, 삭제버튼 추가
  addDeleteBtn();
  // 삭제버튼 추가 후, 마우스 액션 추가
  let deleteBtn = document.querySelectorAll('#result_data tr:first-child > td');
  if(deleteBtn.length > 0) {
    for(let i=1; i<deleteBtn.length; i++) {
      deleteBtn[i].addEventListener("mouseover", columnHover);
      deleteBtn[i].children[0].addEventListener("mouseover", columnHover);
      deleteBtn[i].addEventListener("mouseout", removeStyle);
    }
  }
}

function addDeleteBtn() {
  tbody[0].insertBefore(document.createElement('tr'), tbody[0].firstChild);
  let trTags = document.getElementsByTagName('tr');
  let colCnt = trTags[1].childElementCount;
    for(let i=0; i<colCnt; i++) {
      trTags[0].insertBefore(document.createElement('td'), trTags[0].firstChild);
    }
    for(let i=0; i<colCnt; i++) {
      document.querySelector('#result_data>tbody>tr').children[i].innerHTML = '<i class="far fa-trash-alt"></i>';
      document.querySelector('#result_data>tbody>tr').children[i].classList.add('deleteCol');
    }
    for(let i=0; i<trTags.length; i++) {
      let tdTags =trTags[i].getElementsByTagName('td');
      trTags[i].insertBefore(document.createElement('td'), tdTags[0]);
      if(i !=0){
        tdTags[0].classList.add('deleteRow');
        tdTags[0].innerHTML = '<i class="far fa-trash-alt"></i>';
      }
    }
  addListener();
}

//데이터 추가버튼 클릭시 현재 데이터를 저장한 후 window.close()
function add(){
  let scrapeData = {
    xpathData: xpathArr,
    textData: textArr,
    originData:[],
    originTextData:[],
    secondElementData : [],
    image : []
  };

  chrome.storage.local.set({
    scrapeData : scrapeData
  });

  window.close();
}

//두번째 엘리먼트 선택 시 두번째 엘리먼트 선택 키워드와 이전 스크래핑 내역을 저장 후 window.close()
function addSecondEle(){
  if(typeArr == 'TABLE'){
    alert('테이블은 두번째 엘리먼트를 추가할 수 없습니다.');
    return;
  }
  secondElementArr.push("second");
console.log(originTextArr);
  let scrapeData = {
    xpathData: originArr,
    textData: originTextArr,
    secondElementData : secondElementArr
  };
  
  chrome.storage.local.set({
    scrapeData : scrapeData
  });
console.log(scrapeData);
  setTimeout(() => {
   window.close();
   }, 500);
}

function save(){
  let row = xpathArr.length;
  let col = xpathArr[0].length;
  if(confirm('저장하시겠습니까?')){
    chrome.storage.local.set({
      scrapeData : []
    });
    console.log(image);
    chrome.runtime.sendMessage({xpathList:xpathArr ,textList:textArr, type:'readTable', key:[col,row], image:image});
    window.close();
  }
}

function deleteCol(e) {

  let ele = e.target;
  let table = document.getElementById('result_data');
  let idx = 0;
  if(e.target.tagName == 'I') {
    ele = ele.parentNode;
  }
    while((ele = ele.previousSibling) != null ) {
      idx++;
    }
    for(let i = 0; i < table.rows.length; i++)  {
      table.rows[i].deleteCell(idx);
    }
    for(let i = 0; i < xpathArr.length; i ++){
      xpathArr[i].splice(idx-1,1);
      textArr[i].splice(idx-1,1);
    }
}

function deleteRow(e) {

  let ele = e.target.parentNode;
  let table = document.getElementById('result_data');
  let idx = -1;
  if(e.target.tagName == 'I') {
    ele = ele.parentNode;
  }
    while((ele = ele.previousSibling) != null ) {
      idx++;      
    }
      xpathArr.splice(idx-1,1);
      textArr.splice(idx-1,1);
      table.deleteRow(idx);
}

function addListener(){
  let deleteBtn =  document.querySelectorAll('.deleteCol');
  for(var i = 0  ; i < deleteBtn.length; i++){
    deleteBtn[i].addEventListener('click', deleteCol);
  }
  deleteBtn =  document.querySelectorAll('.deleteRow');
  for(var i = 0  ; i < deleteBtn.length; i++){
    deleteBtn[i].addEventListener('click', deleteRow);
  }
}

function resetTable(){
  if(confirm('데이터를 초기화 하시겠습니까?')){
    chrome.storage.local.set({
      scrapeData : []
    });
    setTimeout(() => {
      window.close();
    }, 500);
  }
}

function columnHover(e) {
  let table = document.getElementById("result_data");
  let ele = e.target;
  let idx = 0; // 현재 선택한 컬럼 index
  
  // 내부 휴지통 아이콘일 경우 현재 Elem을 부모노드로 변경
  if(e.target.tagName == 'I') {
    ele = ele.parentNode;
  }
  while((ele = ele.previousSibling) != null ) {
    idx++;
  }
  if(idx > 0) {
    let select = document.querySelectorAll('tr td:nth-child('+(idx+1)+')');
    for(let i=0; i<select.length; i++) {
      select[i].classList.add('selectCol');
    }     
  }
}

function removeStyle(e) {
  let table = document.getElementById("result_data");
  let ele = e.target;
  let idx = 0; // 현재 선택한 컬럼 index
  while((ele = ele.previousSibling) != null ) {
    idx++;
  }
  if(idx > 0) {
    let select = document.querySelectorAll('tr td:nth-child('+(idx+1)+')');
    select[0].children[0].classList.remove('selectCol');
    for(let i=0; i<select.length; i++) {
      select[i].classList.remove('selectCol');
    }
  }
}

function nextPage(){
    let row = xpathArr.length;
    let col = xpathArr[0].length;
    var comment = '스크래핑할 페이지수를 지정해 주세요. \r\n';
    var pageCnt = window.prompt(comment);
    
    if(pageCnt){
      comment = '웹 페이지에서 다음페이지로 이동하는 버튼을 클릭해 주세요. \r* 숫자가 아닌 페이지 이동버튼을 눌러주세요.';
    }
    if(alert(comment)){
        chrome.storage.local.set({
          scrapeData : []
        });
    }
    chrome.runtime.sendMessage({xpathList:xpathArr ,textList:textArr, type:'selectNextPage', key:[col,row], image:image, pageCnt : pageCnt});
    window.close();

}

document.addEventListener('DOMContentLoaded', getData);

public/background.js
const browserAppData = this.browser || this.chrome;
const tabs = {};
const inspectFile = "inspect.js";
const jqueryFile = "jquery.js";
const activeIcon = "EdenXpath_active.png";
const defaultIcon = "EdenXpath_default.png";
let popupId = "";
let tabList = [];
let lastXpath;
let lastType;
let chrome_storage_lock = false;
let image_save_lock = false;
let nextPageData = {};
let recList = [];
let removeRecList = false; //backspace를 눌러 지운 키보드입력인지 확인하는 변수

const storage = browserAppData.storage && browserAppData.storage.local;

chrome.storage.local.set({
  //처음 로드될 때 모든 데이터 초기화
  data: [],
  scrapeData: [],
});

let started = false;
let tabsUrl = {};

const inspect = {
  toggleActivate: (id, type, icon, url) => {
    started = type === "activate" ? true : false;
    this.id = id;

    browserAppData.tabs.executeScript(null, { file: jqueryFile }, function () {
      browserAppData.tabs.executeScript(id, { file: inspectFile }, function () {
        removeDatascrape();

        if (id == popupId) {
          browserAppData.tabs.sendMessage(id, { action: type });
        } else {
          browserAppData.tabs.sendMessage(id, {
            action: type,
            recList: recList,
          });
        }
      });
    });
    browserAppData.browserAction.setIcon({
      tabId: id,
      path: { 19: "icons/" + icon },
    });
    tabsUrl[id] = url;
  },
};

/**
 * chrome extension을 눌렀을 때 아이콘상태변경 및 contextmenu를 생성
 *
 * @param {*} tab
 */
function toggle(tab) {
  console.log("ok");
  if (tab.title.includes("Worktronics") || tab.title.includes("WORKTRONICS")) {
    alert(
      "디자인캔버스에서는 웹레코딩을 실행할 수 없습니다.\n다른탭으로 이동한 후 웹레코딩을 실행해주세요."
    );
  }
  if (tab.status === "complete") {
    chrome.tabs.setZoom(1);
    var wdTab = !(
      tab.title === undefined ||
      tab.title.includes("Worktronics") ||
      tab.title.includes("WORKTRONICS")
    );
    if (isSupportedProtocolAndFileType(tab.url) && wdTab) {
      if (!tabs[tab.id]) {
        tabs[tab.id] = Object.create(inspect);
        inspect.toggleActivate(tab.id, "activate", activeIcon, tab.url);
        chrome.contextMenus.create({
          id: "readText",
          title: "텍스트 읽기",
          contexts: ["all"],
          onclick: contextClick,
        });
        chrome.contextMenus.create({
          id: "dataScrapping",
          title: "데이터스크래핑",
          contexts: ["all"],
          onclick: contextClick,
        });
        chrome.contextMenus.create({
          id: "readAttribute",
          title: "속성값 읽기",
          contexts: ["all"],
          onclick: contextClick,
        });
        chrome.contextMenus.create({
          id: "href",
          title: "href",
          contexts: ["all"],
          parentId: "readAttribute",
          onclick: contextClick,
        });
        chrome.contextMenus.create({
          id: "value",
          title: "value",
          contexts: ["all"],
          parentId: "readAttribute",
          onclick: contextClick,
        });
        chrome.contextMenus.create({
          id: "class",
          title: "class",
          contexts: ["all"],
          parentId: "readAttribute",
          onclick: contextClick,
        });
        chrome.contextMenus.create({
          id: "hover",
          title: "hover 클릭",
          contexts: ["all"],
          onclick: contextClick,
        });

        let data = [
          {
            //웹레코더 실행한 페이지를 레코딩
            url: tab.url || tab.pendingUrl,
            type: "newTab",
            xpath: "",
            key: "",
          },
        ];

        addAction(data);
        checkTab("start", tab.id, tab.index); // 시작할 때 탭리스트를 담아서 이후 꺼지거나 생성되는 탭을 구별함
      } else {
        chrome.contextMenus.removeAll();
        tabList = [];
        browserAppData.tabs.query({}, (tabs) => {
          tabs.forEach((t) => {
            if (
              !(t.title === undefined) &&
              !t.title.includes("Worktronics") &&
              !t.title.includes("WORKTRONICS")
            ) {
              recList = [];
              inspect.toggleActivate(t.id, "deactivate", defaultIcon, tab.url);
            }
          });
        });

        for (const tabId in tabs) {
          delete tabs[tabId];
          if (tabId == tab.id) {
            browserAppData.tabs.sendMessage(tab.id, {
              action: "record",
              tabId: tab.id,
            });
          }
        }
      }
    }
  }
}

/**
 * contextmenu 하위메뉴를 클릭했을 때, 어떤메뉴를 클릭했을지 알수없음으로 클릭하는 시점에 어떤메뉴를 클릭했는지 메세지를 보내줌
 *
 * @param {*} info
 * @param {*} tab
 */
function contextClick(info, tab) {
  var req = { cmd: info.menuItemId, url: tab.url, command: "contextMenu" };
  browserAppData.tabs.sendMessage(tab.id, req);
}

/**
 * activetab에 대한 속성을 읽어와서 지원되는 형식의 사이트인지 확인
 *
 * @param {*} urlString
 * @returns
 */
function isSupportedProtocolAndFileType(urlString) {
  if (!urlString) {
    return false;
  }
  const supportedProtocols = ["https:", "http:", "file:"];
  const notSupportedFiles = ["xml", "pdf", "rss"];
  const extension = urlString.split(".").pop().split(/\#|\?/)[0];
  const url = document.createElement("a");
  url.href = urlString;
  return (
    supportedProtocols.indexOf(url.protocol) !== -1 &&
    notSupportedFiles.indexOf(extension) === -1
  );
}

/**
 * 새탭이 생겼을 때 몇번째 탭이 생기고 닫혔는지 확인해 탭의 위치를 리턴
 *
 * @param {*} type
 * @param {*} tbId
 * @param {*} tbIdx
 * @returns
 */
function checkTab(type, tbId, tbIdx) {
  var idx;
  if (type == "start") {
    tempList = { tbId, tbIdx };
    tabList.push(tempList);
  } else if (type == "new") {
    tempList = { tbId, tbIdx };
    tabList.push(tempList);
    idx = tabList.length - 1;
  } else if (type == "close") {
    for (var i = 0; i < tabList.length; i++) {
      if (tabList[i].tbId == tbId) {
        tabList.splice(i, 1);
        idx = i - 1;
        break;
      }
    }
  }
  return idx;
}

function getActiveTab(tabId, changeInfo, tab) {
  console.log("get");
  if (changeInfo.favIconUrl) {
  }
  if (changeInfo.status === "complete" && tab.title != undefined) {
    if (
      started &&
      tab.title &&
      !(tab.title.includes("Worktronics") || tab.title.includes("WORKTRONICS"))
    ) {
      tabs[tab.id] = Object.create(inspect);
      inspect.toggleActivate(tab.id, "activate", activeIcon, tab.url);
      //   browserAppData.tabs.sendMessage(tabId, {
      //     action: 'addViewDiv'
      //    ,recList : recList
      //  });
    }
  }
}

browserAppData.commands.onCommand.addListener((command) => {
  if (command === "toggle-xpath") {
    browserAppData.tabs.query({ active: true, currentWindow: true }, (tab) => {
      toggle(tab[0]);
    });
  }
});

/**
 * Message handler function for adding a action to action list.
 * Action list should be solely managed by background.js, content.js should not handle chrome.storage directly.
 * Because of handling multiple tabs safely & avoiding race conditions for chrome.storage.
 *
 * @param {*} action
 * @returns
 */
function addAction(action) {
  if (chrome_storage_lock == true) {
    // wait 1 ms
    setTimeout(addAction, 1, action);
    return;
  }

  if (image_save_lock == true) {
    // wait 5 ms
    setTimeout(addAction, 10, action);
    return;
  }

  chrome_storage_lock = true; // get chrome storage lock

  storage.get(
    {
      data: [],
    },
    function (result) {
      if (result.data.length !== 0) {
        let actionList = result.data;
        if (lastType === "none") {
          actionList.pop();
        } else if (
          action[0].type === "keyboard" &&
          action[0].type === lastType &&
          action[0].xpath === lastXpath &&
          action[0].key.length != 0
        ) {
          action[0].image = actionList[actionList.length - 1][0].image;
          actionList.pop();
          recList.pop();
        } else if (
          lastType === "Enter" &&
          action[0].xpath === lastXpath &&
          action[0].type != "keyboard"
        ) {
          actionList.pop();
          recList.pop();
        } else if (
          action[0].type === "hover" &&
          action[0].type === lastType &&
          action[0].xpath === lastXpath
        ) {
          actionList.pop();
          recList.pop();
        }

        if (
          !removeRecList &&
          !(action[0].type === "keyboard" && action[0].key.length == 0)
        ) {
          actionList.push(action);
        }

        removeRecList = false; // removeRecList 초기화
        if (action[0].rectext != "" && action[0].rectext != null) {
          recList.push(action[0].rectext);
        }
        storage.set({
          data: actionList,
        });
      } else {
        let actionList = [];
        actionList.push(action);

        storage.set({
          data: actionList,
        });
      }
      lastXpath = action[0].xpath;
      lastType = action[0].type;
      chrome_storage_lock = false;
      // release chrome storage lock
    }
  );
}

/**
 * Message handler for requests from multiple content.js tabs.
 *
 * @param {*} request
 * @param {*} sender
 * @param {*} sendResponse
 */
async function recMsg(request, sender, sendResponse) {
  if (request.type == "readTable") {
    removeDatascrape(); //기존 데이터스크래핑내역 초기화
    let data = [
      {
        url: "noPage",
        type: "dataScrapping",
        xpath: request.xpathList,
        key: request.key,
        image: request.image,
      },
    ];
    addAction(data);
  } else if (request.type == "readTableNextPage") {
    removeDatascrape(); //기존 데이터스크래핑내역 초기화
    let pageCnt = nextPageData.pageCnt;
    let pageXpath = request.pageXpath;
    let nextPageList = [pageCnt, pageXpath];
    let data = [
      {
        url: nextPageList,
        type: "dataScrapping",
        xpath: nextPageData.xpathList,
        key: nextPageData.key,
        image: nextPageData.image,
      },
    ];
    addAction(data);
  } else if (request.type == "addAction") {
    if (
      !(
        request.data[0].type === "keyboard" &&
        request.data[0].type === lastType &&
        request.data[0].xpath === lastXpath
      ) &&
      request.coords
    ) {
      request.data[0].image = await capture(request.coords, "normal");
    }
    addAction(request.data);
  } else if (request.type == "createPopup") {
    let image = await capture(request.coords, "scrape");
    request.data[0].image = image;
    chrome.storage.local.set(
      {
        scrapeData: request.data,
      },
      () => {
        chrome.windows.create({
          url: chrome.runtime.getURL("data.html"),
          type: "popup",
          height: 500,
          width: 800,
        });
      }
    );
  } else if (request.type == "recordComplete") {
    browserAppData.tabs.query({}, (tabs2) => {
      tabs2.forEach((tab) => {
        if (tab.title !== undefined) {
          if (
            tab.title.includes("Worktronics") ||
            tab.title.includes("WORKTRONICS")
          ) {
            browserAppData.tabs.sendMessage(tab.id, {
              action: "createRecord",
              data: request.data,
            });
            delete tabs[request.tabId];
          }
        }
      });
    }); // tabs.query 끝.
  } else if (request.type == "selectNextPage") {
    nextPageData = request;
    var last = tabList.length - 1;

    browserAppData.tabs.sendMessage(tabList[last].tbId, {
      action: "selectNextPage",
      pageCnt: nextPageData.pageCnt,
    });
  } else if (request.type == "removeRecList") {
    removeRecList = true;
    lastType = "removeRecList";
    recList.pop();
  }
}

/**
 *
 * @param {*} coords //엘리먼트 좌표
 * @param {*} type //데이터스크래핑 || others
 * @returns
 */
function capture(coords, type) {
  if (image_save_lock == true) {
    // wait 1 ms
    setTimeout(capture, 1, capture);
    return;
  }
  image_save_lock = true; // get chrome storage lock

  return new Promise(function (resolve, reject) {
    chrome.tabs.captureVisibleTab(null, { format: "jpeg" }, function (dataUrl) {
      if (chrome.runtime.lastError) {
        image_save_lock = false;
        resolve("error");
      } else {
        var img = new Image();

        img.onload = function () {
          var imgCanvas = document.createElement("canvas");

          if (coords.x + coords.w > img.width) {
            coords.w = img.width - coords.x;
          }
          if (coords.y + coords.h > img.height) {
            coords.h = img.height - coords.y;
          }
          //220315 그림제외버전
          // imgCanvas.width = coords.w;
          // imgCanvas.height = coords.h;
          // var ctx_ = imgCanvas.getContext('2d');
          // ctx_.fillStyle = 'rgba(0,0,0,0.5)';
          // ctx_.drawImage(img, coords.x, coords.y, coords.w, coords.h, 0, 0, coords.w, coords.h);
          // ctx_.fillStyle = 'rgba(0,0,0,0.5)';
          // imageData = imgCanvas.toDataURL('image/jpeg', 1.0);

          //그림그리는 버전
          imgCanvas.width = img.width;
          imgCanvas.height = img.height;

          var ctx = imgCanvas.getContext("2d");
          ctx.drawImage(img, 0, 0, img.width, img.height);
          imageData = imgCanvas.toDataURL("image/jpeg", 1.0);
          if (type == "normal") {
            ctx.lineWidth = 5;
            ctx.strokeStyle = "rgba(255,108,108)";
            ctx.strokeRect(coords.x, coords.y, coords.w - 200, coords.h - 200);
          }
          imageData = imgCanvas.toDataURL("image/jpeg", 1.0);

          var drawImg = new Image();
          drawImg.src = imageData;
          drawImg.onload = function () {
            var canvas = document.createElement("canvas");
            canvas.width = coords.w;
            canvas.height = coords.h;

            var ctx_ = canvas.getContext("2d");
            ctx_.fillStyle = "#fff";
            ctx_.fillRect(0, 0, canvas.width, canvas.height);

            if (type == "scrape") {
              ctx_.drawImage(
                drawImg,
                coords.x,
                coords.y,
                coords.w,
                coords.h,
                0,
                0,
                coords.w,
                coords.h
              );
            } else {
              ctx_.drawImage(
                drawImg,
                coords.x - 100,
                coords.y - 100,
                coords.w,
                coords.h,
                0,
                0,
                coords.w,
                coords.h
              );
            }

            imageData = canvas.toDataURL("image/jpeg", 1.0);
            image_save_lock = false;
            resolve(imageData);
          };
        };
        img.src = dataUrl;
      }
    });
  });
}

function removeTab(tabId, tab) {
  let data = [];
  console.dir(tabList);
  if (started) {
    if (tabId == popupId) {
      data = [
        {
          url: tabsUrl[tabId],
          type: "closePopup",
          xpath: "",
          key: "",
        },
      ];
    } else {
      var idx = checkTab("close", tabId, "");
      if (idx != undefined && idx > -1) {
        data = [
          {
            url: tabsUrl[tabId],
            type: "moveTab",
            xpath: "",
            key: idx,
          },
        ];
      } else {
        return;
      }
    }
    addAction(data);
  }
}

function movedTab(tabId, moveInfo) {
  if (started) {
    let data = [
      {
        url: tabsUrl[tabId],
        type: "closeTab",
        xpath: "",
        key: "",
      },
    ];

    addAction(data);
  }
}

function createTab(tab) {
  if (started) {
    console.log("createTab");
    let data = [];
    browserAppData.tabs.query({ active: true, currentWindow: true }, (t) => {
      console.log(t[0]);
      if (
        t[0].pendingUrl != null &&
        t[0].pendingUrl.includes("chrome-extension://")
      ) {
        return;
      }
      console.log("popup");
      if (t[0].pendingUrl != undefined && t[0].incognito != true) {
        let tempUrl = t[0].pendingUrl;
        if (t[0].index == 0) {
          console.log("pop업!");
          popupId = t[0].id;
          data = [
            {
              url: tempUrl,
              type: "popup",
              xpath: "",
              key: "",
            },
          ];
        } else {
          var idx = checkTab("new", t[0].id, t[0].index);
          data = [
            {
              url: tempUrl,
              type: "moveTab",
              xpath: "",
              key: idx,
            },
          ];
        }
        addAction(data);
      }
    });
  }
}

function removeDatascrape() {
  // 데이터스크래핑내역 초기화
  chrome.storage.local.set({
    xpathData: [],
    textData: [],
    typeArr: [],
    originData: [],
    originTextData: [],
    secondElementData: [],
  });
}

//manifest.json의 content_scripts에 맞는 url 일때 inspect.js 삽입
chrome.webNavigation.onHistoryStateUpdated.addListener(function (details) {
  chrome.tabs.executeScript(null, { file: inspectFile });
});

//storage changes 감지하는 부분
// chrome.storage.onChanged.addListener(function(changes, namespace) {
//   console.log('cahnges;');
//   console.log(changes);
// });
const popupToggle = () => {
  chrome.windows.create({
    url: chrome.runtime.getURL("index.html"),
    type: "popup",
    height: 600,
    width: 1000,
  });
};
const toggleMessage = (request, sender, sendResponse) => {
  if (request.type == "toggleActivate") {
    let queryOptions = { active: true };
    chrome.tabs.query(queryOptions, ([tab]) => {
      if (chrome.runtime.lastError) console.error(chrome.runtime.lastError);
      // `tab` will either be a `tabs.Tab` instance or `undefined`.

      toggle(tab);
    });
  }
};
browserAppData.tabs.onUpdated.addListener(getActiveTab);
browserAppData.tabs.onRemoved.addListener(removeTab);
browserAppData.tabs.onCreated.addListener(createTab);
browserAppData.tabs.onMoved.addListener(movedTab);
browserAppData.browserAction.onClicked.addListener(popupToggle);
browserAppData.runtime.onMessage.addListener(recMsg);
browserAppData.runtime.onMessage.addListener(toggleMessage);
